### ISSUE TITLE
Rework `Iterator`

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#5249]
I would like to have `Iterator` re-implemented. (Probably almost from scratch).

The current type definition is:

```
pub enum Iterator[a: Type, r: Region] {
    case Iterator(Iterator.Done[r], Iterator.Next[a, r])
}
```

This should instead become:

```
pub enum Iterator[a: Type, r: Region] {
    case Iterator(Region[r], Unit -> Option[a] \ r)
}
```

Thus the `Done` and `Next` functions are merged into one. 
And `Iterator` now carries its own `Region[r]` (hence we can get rid of `Scoped`).

In addition to these (fundamental) changes, I would like:

- Every operation to be lazy. (Thus dropping the functions that currently return `List`).

Undecided:
- What should the signature of e.g. `append` be?

**Comment by stephentetley** [comment]
I've a working implementation here - temporarily called `Cursor` so it doesn't clash with the current `Iterator`.

https://github.com/stephentetley/flix-doodle/blob/master/temp/Cursor.flix

Much of the original Iterator API is covered. `intersperse` turns out very nicely when written in terms of `flatMap`, trying to write it directly without a test for "done" is quite a knotty problem. I use the `Step` datatype internally rather than `Option` this makes the implementation of `filter` easy. All the functions except `toList` and the fold related ones should be "streamable" / "steppable" (roughly speaking "lazy"), `intercalate` is an outlier as it has to copy the separating iterator each time it is used, I think the original `Iterator` was doing this as well.

I haven't seriously looked at Regions because it's complicated. The answer is probably to follow `Array` and pass a region to hold the answer as a parameter, but to make the current implementation simple I've tried to keep to the signatures of the original implementation.

**Comment by stephentetley** [comment]
I've made a new version with "output regions" as the first parameter of the function signatures as per the current version of Array in the stdlib.

https://github.com/stephentetley/flix-doodle/blob/master/temp/Cursor2.flix

**Comment by magnus-madsen** [comment]
@stephentetley I would like you adopt your version `Cursor` as the new `Iterator`. I think it looks clean with `Step` :)

My major concern is all the casts. Take for example:

```
    pub def map(r1: Region[r1], f: a -> b \ ef, iter: Cursor[a, r]): Cursor[b, r1] \ { ef, Read(r), Write(r1) } =
        let Cursor(_, iterF) = iter;
        let step = x -> match x {
            case Ans(a) => Ans(f(a))
            case Skip   => Skip
            case Done   => Done
        };
        let iterF1 = () -> unsafe_cast step(iterF()) as _ \ r1;
        unsafe_cast Cursor(r1, iterF1) as _ \ { ef, Read(r), Write(r1) }
```

Since map is lazy, I would expect it to have this simple signature:

```
    pub def map(f: a -> b \ ef, iter: Cursor[a, r]): Cursor[b, r and ef] = ...
```

As for "output regions", I think that is only relevant when creating a whole new mutable data structure for scratch. I think the right abstraction is that `Iterator` (exactly like `MutList`) stores a reference it its region and allocate all new mutable data in that region. Does that seem sensible? It should give reasonable signatures like the above.

**Comment by stephentetley** [comment]
Thanks Magnus, I'll work up the first version of Cursor to be a replacement for Iterator. I'm back at work now after the Christmas holidays so I'll be a bit slower than I have just been.

**Comment by magnus-madsen** [comment]
Perhaps draft a few functions and see if they work without casts and then ping me :)

**Comment by stephentetley** [comment]
I've looked at eliminating casting.

The ideal implementation of `Iterator` fails for `map` - the use of `f` within `step` wher the case is `Ans` makes the iterator function `Unit -> Step[a] \ {ef, r}` rather than keeps it at `Unit -> Step[a] \ r`


~~~

pub enum Iterator[a: Type, r: Region] {
    case Iterator(Region[r], Unit -> Step[a] \ r)
}

enum Step[a] {
    case Ans(a)
    case Skip
    case Done
}


/// Err...
pub def map(f: a -> b \ ef, iter: Iterator[a, r]): Iterator[b, r] \ ef =
    let Iterator(r, iterF) = iter;
    let step = x -> match x {
        case Ans(a) => Ans(f(a))
        case Skip   => Skip
        case Done   => Done
    };
    let iterF1 = () -> {let x = iterF(); step(x)};
    Iterator(r, iterF1)


~~~

Adding `ef` as a type param allows us to write `map` without casting:

~~~

pub enum Iterator[a: Type, ef: Bool, r: Region] {
    case Iterator(Region[r], Unit -> Step[a] \ {r, ef})
}


pub def map(f: a -> b \ ef1, iter: Iterator[a, ef, r]): Iterator[b, ef and ef1, r] =
    let Iterator(r, iterF) = iter;
    let step = x -> match x {
        case Ans(a) => Ans(f(a))
        case Skip   => Skip
        case Done   => Done
    };
    let iterF1 = () -> {let x = iterF(); step(x)};
    Iterator(r, iterF1)

~~~

Unfortunately we can't have "unencapsulated loops" any more. The original `flatMap` has an unencapsulated loop - it is not returned within an `Iterator` constructor, the return type of loop is an `Iterator` rather than a `Step` and that `Iterator` gets returned by the function:

~~~

/// Type err...
pub def flatMap(f: a -> Iterator[b, ef1, r] \ ef2, ma: Iterator[a, ef, r]): Iterator[b, ef and ef1 and ef2, r] =
    let Cursor(r, iterF) = ma;
    def loop(x, acc) = match x {
        case Ans(a) => loop(iterF(), append(acc, f(a)))
        case Skip   => loop(iterF(), acc)
        case Done   => acc
    };
    loop(iterF(), empty(r))

~~~

I think `flatMap` should be implementable with an encapsulated loop but I haven't managed it yet.

The code so far is here with `Iterator` again renamed to `Cursor` 

https://github.com/stephentetley/flix-doodle/blob/master/temp/Cursor3.flix

**Comment by magnus-madsen** [comment]
Thanks for the examples Stephen. I understand the problems now. Let me see if I can come up with something.

**Comment by magnus-madsen** [comment]
I looked over `Cursor3.flix`. This is great work-- as usual 👍. I cannot offer any new insights yet, but I have some comments:

- The switch to `enum Iterator[a: Type, ef: Bool, r: Region]` is sensible and reasonable. 
- I notice that `Cursor3` seems to contain very few casts. If we only need a handful of casts (e.g. in `flatMap` that is okay with me).
- I believe some iterator implementations uses a `Step` data type that also has a `FlatMap` constructor. Do you think that could be helpful here? (I admit to not yet knowing how it would look).
- I noticed that VSCode complains about several complex types with large Boolean formulas. That's interesting and something I will look into. (This is not a fault of the code, but rather a suggestion that there is more stuff to do on the type inference side).

Overall, I think we should just continue with the `Cursor3.flix` approach.

**Comment by stephentetley** [comment]
I've got `flatMap` working on Cursor3 with no casts and keeping the definition of `Cursor` the same. I'm also getting quite near having an equivalent API to the current Iterator in the stdlib.

> I believe some iterator implementations uses a Step data type that also has a `FlatMap` constructor

I think they would probably be written in the "operational" / free monad style. The downside is all `match` uses have to consider the `FlatMap` case so I was pleased I could write Cursor3 `flatMap` without it.

I expect the experience writing this `flatMap` will help writing "functional monads". Before Christmas I put aside working on my "MonadLib" library because I was disappointed with the large amount of casts it seemed to need, this work on Iterators should give me some fresh ideas to apply to it.

**Comment by magnus-madsen** [comment]
> I've got flatMap working on Cursor3 with no casts and keeping the definition of Cursor the same. I'm also getting quite near having an equivalent API to the current Iterator in the stdlib.

Note: I am okay with dropping a few functions. For example, I think there should only be one `filter` which should be the lazy one. Not two like today where one returns an iterator.

**Comment by stephentetley** [comment]
I've written `intercalate` with a cusror-of-cursors so I think I've pretty much covered the existing API.

I'll start work on a branch where I replace the existing `Iterator` at the weekend.

I had to change to use a few more casts - I think `new`, `singleton` and the other region "creators" should be effect polymorphic otherwise they are difficult to use in the definition of other functions `intersperse`. Maybe we need effect "lifters" similar to lifting between monadic layers in Haskell, but that can be investigated at another time.

**Comment by magnus-madsen** [comment]
> I've written `intercalate` with a cursor-of-cursors so I think I've pretty much covered the existing API.

Great news! 

> I'll start work on a branch where I replace the existing `Iterator` at the weekend.

👍 

> I had to change to use a few more casts - I think `new`, `singleton` and the other region "creators" should be effect polymorphic otherwise they are difficult to use in the definition of other functions like `intersperse`. Maybe we need effect "lifters" similar to lifting between monadic layers in Haskell to avoid casts, but that can be investigated at another time.

OK. I could you provide an example with a bit more detail?

Finally, I think we still need some kind of zipper on RedBlackTree, is that right?

**Comment by paulbutcher** [comment]
I've been working on removing `Scoped.regionOf` as part of #5131. Iterator is the only remaining place where it's still used. I'm assuming that it's best for me to hold off doing anything until this ticket is complete? But let me know if not.

**Comment by magnus-madsen** [comment]
Yes. I think we will completely replace it.

**Comment by stephentetley** [comment]
I've now re-implemented the API, and reworked the tests but the types are very challenging and the effect typing doesn't feel right. For instance, some of the functions with two input iterators I've had to restrict to both having to use the same region. I think this is a necessity for `flatMap` but shouldn't be for others. 

Shall I make a "WIP" PR anyway so at least the work is visible?

> Finally, I think we still need some kind of zipper on RedBlackTree, is that right?

I think I used the existing `iterator` code as is, just removing the `done` function.
